numpyro/util.py

```diff
@@ -201,24 +202,40 @@ def progress_bar_factory(num_samples, num_chains):
 
     remainder = num_samples % print_rate
 
+    idx_map = {}
     tqdm_bars = {}
     finished_chains = []
+    lock = Lock()
```

Comment:
could you add a comment for why this is needed?

Reply:
done

Reply:
Could you clarify the purpose of Lock in more details? I'm not familiar with its usage. What happens if we dont use it?

Reply:
Updated.  The locking is only around `idx_counter` since after a chain has a chain id there isn't access to resources that are shared across threads.